### ISSUE TITLE
Question Regarding static vs dynamic typing

### ISSUE THREAD

**Issue by Mojo90** [sylvainpolletvillard/ObjectModel#46]
I saw in your Description that you say `These solutions bring static typing, which means it only validate your code at build time, not runtime. Once compiled in JavaScript and run in the browser, there is no longer any guarantee that the variables you are working with have the intended types.` As I can see from [TypeScript Docs](https://www.typescriptlang.org/docs/handbook/advanced-types.html) the `Advanced Types - User-Defined Type Guards` are exactly the case to detect the correct type of a variable during runtime, isn't it?

**Comment by sylvainpolletvillard** [comment]
Not really, no. This feature is meant to give indications to TypeScript compiler that when you call a function at runtime like `isFish` (a function that you have to write yourself), you bring the guarantee to TypeScript that the code under this `if()` branch has the `Fish` type, and TS can narrow down the union type.

TypeScript does not provide runtime checking, but the developer does and then has to inform TypeScript that he has done the job :smile:
